### DISCUSSION TITLE
How to simply iterate over nodes/elements to compute a source term used in a Kernel?

### DISCUSSION THREAD

**Discussion by lgab13** [idaholab/moose#18675] [Q&A General] [answered]
Hi Moose experts,

 How iterate over the elements/nodes of the mesh to compute, at the beginning or at the end of each time step, a  source term that depends on one or more fields and that is used in a Kernel? 

Please consider the following (absurd) example

```C++

for(unsigned_int n=0; n<nnodes;n++){

   MySourceTerm[n] = sqrt(Temperature[n]) ;   

}

```

What is the best way to do it (AuxKernels, Kernels, UserObject,...) ? 

Do you have any examples?

It is likely that this loop should be performed at a higher level but I did not find these explanations in the documentation. 
Is there a document describing a generic "solving tree" with the name of the methods and the order in which they are called at runtime:

preprocessing()?
initialize() ?
execute() ?
postprocessing()?
finalize()?

Best regards,

lg

**Comment by GiudGiud** [comment, marked as answer]
Hello

I would add an AuxVariable `source`, to the simulation, and a new AuxKernel that computes the custom source.
If the expression of source is simple, like sqrt(T), you can use a [ParsedAux](https://mooseframework.inl.gov/source/auxkernels/ParsedAux.html) for this purpose

For specifying beginning/end of time step for computing the new source, you can use the `execute_on = TIMESTEP_BEGIN / END` parameter of the AuxKernel. 

The Kernel that uses the source in the equation can then take it as a CoupledVar parameter. Note that if this kernel is a simple `constant * variable` term there are already pre-implemented kernels for this such as [CoupledForce](https://mooseframework.inl.gov/source/kernels/CoupledForce.html)

Best,
Guillaume
